internal/fnruntime/container_test.go

```diff
@@ -69,3 +70,39 @@ func TestContainerFn(t *testing.T) {
 		})
 	}
 }
+
+func TestFnImagePull(t *testing.T) {
+	image := "gcr.io/kpt-fn/search-replace:v0.1"
+
+	// Initial test setup (our environment must not already include the image)
+	cmdInspect := exec.Command("docker", "image", "inspect", image)
+	errInspect := cmdInspect.Run()
+	// docker image inspect exits successfully if an image is found
+	// If image is already present remove the image before running test
+	if errInspect == nil {
+		cmdRm := exec.Command("docker", "image", "rm", image)
+		errRm := cmdRm.Run()
+		// Image should remove successfully
+		assert.NoError(t, errRm)
+	}
+	// Test setup complete
+
+	// Exec function (this should pull the missing function image)
+	ctx := context.Background()
+	outBuff, errBuff := &bytes.Buffer{}, &bytes.Buffer{}
+	instance := fnruntime.ContainerFn{
+		Ctx:   printer.WithContext(ctx, printer.New(outBuff, errBuff)),
+		Image: image,
+	}
+	// Function execution should be a no-op
+	input := bytes.NewBufferString("")
+	output := &bytes.Buffer{}
+	err := instance.Run(input, output)
+	assert.NoError(t, err)
+
+	// Inspect fails with an exit code of 1 if no image is present
+	// We expect this command to succeed (the image exists)
+	cmdImageInspect := exec.Command("docker", "image", "inspect", image)
+	errImageInspect := cmdImageInspect.Run()
+	assert.NoError(t, errImageInspect)
+}
```

Comment:
It's not clear to me what exactly are we testing here ?

Reply:
When you invoke `instance.Run()`:
1. Image could be pulled using the `prepareImage()`
2. Or it could be pulled using the `docker run`

so either case, `docker image inspect` will succeed.. right ? or may be I am missing something.

Reply:
This is specifically targeting the container functions and that executing the container Command will pull the image. It is not testing **how** that image is pulled (yet). This is meant as a proxy for manually testing this functionality.

`docker image inspect` exits with non-zero exit code if the image does not exist. A quick example:

```
$ docker pull postgres:latest
latest: Pulling from library/postgres
Status: Downloaded newer image for postgres:latest
docker.io/library/postgres:latest

$ docker image rm postgres:latest
Untagged: postgres:latest
...

$ docker image inspect postgres:latest
[]
Error: No such image: postgres:latest
```

This test is meant to:

1. Remove any existing local image reference (if one exists)
2. Execute the function as a no-op
3. Validate that the function image is available locally

The test case I had in mind was:

> Given an environment that does not have a function image locally, when kpt executes that function, then the function image is checked out locally.

This could be refactored to target specifically covering `prepareImage` if we'd prefer but is currently broader than that. If you'd like I could append a test to this PR that introduces that coverage.

Reply:
> This could be refactored to target specifically covering prepareImage if we'd prefer but is currently broader than that. If you'd like I could append a test to this PR that introduces that coverage.

No, not in favor of testing internal functions. Infact, I would like the test case to be:

> Given an environment that does not have a function image locally, when kpt _should be able to execute_ that function successfully.

Ideally having it in e2e test case would be better but we don't have a way to setup the test-case today where we ensure the function image isn't present before running the test. 

But this is fine for now.

Reply:
Today @Shell32-Natsu added support for running `setup.sh` script in e2e test framework https://github.com/GoogleContainerTools/kpt/pull/2032

So, we can implement this test as e2e test. That will be better. So copy one of the simple tests and add `setup.sh` that deletes the docker images being used in that test and successful run of the test will achieve the same outcome as here.